### DISCUSSION TITLE
TabBarController의 configureViewControllers(withUser user:)

### DISCUSSION THREAD

**Discussion by z3rosmith** [swift-in-gangnam/soil#23] [Ideas]
지금 구조가 user가 설정되지 않으면 configureViewControllers가 실행되지 않는데 user와 관계없이 configure는 일어나야한다고 봅니다. 수정하는게 어떨까요? 
이렇게되면 youController를 init할때 user를 넘겨주는 구조도 바꿔야 하는데 init할때 꼭 user를 넘겨주지 않아도 되지 않나요? 의견이 어떤지 궁금합니다.

**Comment by Chuncheonian** [comment]
메인으로 들어올려면 User가 있어야되지않나요? 로그인해야만 들어올 수 있으니까..?

**Reply by z3rosmith** [reply]
아 제가 표현을 좀 정확하지 않게 했네요. 메인으로 들어오려면 User는 있어야하는데 fetchUser가 완료되기 전에도 UI의 configure가 일어나야 할것같습니다. 지금은 fetchUser가 완료되기 전에는 UI의 configure가 일어나지 않아서 까만화면에 탭바도 글자가 표시가 안되네요

**Reply by Chuncheonian** [reply]
자동로그인이 되어있는상태에서 실행할 때 한 1초동안 검은색 화면때문에 그러시는건가요?

**Reply by z3rosmith** [reply]
넵 UI초기화랑 fetchUser는 분리되는게 좋아보여요

**Reply by Chuncheonian** [reply]
아직 이해를 못했는데 분리되어있지않나요?
그리고 검은색화면은 아래와 같이 view.backgroundColor = .white만 위치바꾸면 되지 않을까요?

```swift
  override func viewDidLoad() {
    super.viewDidLoad()
    view.backgroundColor = .white
    checkIfUserIsLoggedIn()
    fetchUser()
  }
```

**Reply by Chuncheonian** [reply]
다른 UI는 로그인안되어있으면 굳이 보여줄 필요없지 않나요?

**Reply by z3rosmith** [reply]
코드에서 보시면 fetchUser()에서 self.user가 user로 설정될때 didSet이 일어나면 그때 configure가 일어나는데 이걸 fetchUser와 configure가 분리되어있지 않다고 말씀드린거에요 음 configure가 backgroundColor 말고도 tabBar의 설정도 다 같이 해주고있는데 탭바는 로그인되는중에도 보여야되지않을까요?

```swift
var user: User? {
  didSet {
    guard let user = user else { return }
    configureViewControllers(withUser: user)
  }
}

func fetchUser() {
    UserService.fetchUser { user in
       self.user = user
    }
}
```

**Reply by Chuncheonian** [reply]
흠 그러네요 이상하긴 하네요
<img width="25%" alt="스크린샷 2021-11-18 오후 10 44 36" src="https://user-images.githubusercontent.com/29334434/142426766-be166ca2-155b-4305-bf6f-767f57dadfde.png">

**Reply by z3rosmith** [reply]
configureViewController에서는 UI만 변경하고 이걸 viewDidLoad에 넣은 후에

```swift
var user: User? {
  didSet {
    guard let user = user else { return }
    setYouControllerUser(withUser: user)
  }
}
```

여기서 이런식으로 user 설정해도 좋을거같아요 YouController안에서는 

```swift
// private var user: User
var user: User?
```

이렇게 바꾼 후에 init때 초기화하지 않고 setYouControllerUser(withUser:) 안에서 

```swift
func setYouControllerUser(withUser user: User) {
  ...
  youController.user = user
  ...
}
```

이런식으로 설정하구요

**Reply by Chuncheonian** [reply]
checkIfUserIsLoggedIn()를 AppDelegate나 SceneDelegate에 할까요?

```swift
  func checkIfUserIsLoggedIn() {
    if Auth.auth().currentUser == nil {
      DispatchQueue.main.async {
        let controller = UIStoryboard(name: "Auth", bundle: nil).instantiateViewController(withIdentifier: "LoginVC") 
        //controller.delegate = self
        let nav = UINavigationController(rootViewController: controller)
        nav.modalPresentationStyle = .fullScreen
        self.present(nav, animated: true, completion: nil)
      }
    }
  }
```

**Reply by z3rosmith** [reply]
아 넵 그것도 그렇게 하는게 좋을거같네요 그리고 생각난건데 user를 VC에 저장하고 있을 필요가 없을거같아요

- https://developer.apple.com/documentation/foundation/userdefaults
- https://zeddios.tistory.com/107

userDefaults(비밀번호 외 저장)랑 Keychain(비밀번호저장)을 사용하면 될거같아요 아직 키체인의 경우는 제가 못알아봤는데 UserDefaults는 비밀번호 저장용도로는 취약할수 있어서 키체인을 비밀번호 저장용도로 써라고 하네요

**Reply by z3rosmith** [reply]
혹시 user를 따로 두신 이유가 있으신가요??

**Reply by Chuncheonian** [reply]
Youcontroller말고 FeedController하고 UploadPostController에도 다 User가 필요하지않을까요? 다 옵셔널처리하기엔 코드가 길어질거같아요

**Reply by Chuncheonian** [reply]
SceneDelegate에서 먼저 fetch를하고 Tabbar에 init을 만들어서 user를 전달해주는건 어떨까요?

**Reply by z3rosmith** [reply]
user를 따로 저장하려는 이유가 있으신가요? UserDefaults에 저장해놓고 쓰면 되지않을까요?

**Reply by Chuncheonian** [reply]
아 원래 그런 생각은 했는데 지금은 조금밖에 못해서 나중에 구현할려고 생각했어요
